File: dhandho-core/src/main/java/cc/dhandho/jsclient/ng/CommandComponent.java
package cc.dhandho.jsclient.ng;

import java.util.function.Function;

import def.angular.core.Component;
import def.angular.http.Http;
import def.angular.http.Response;
import def.dom.Globals;
import def.js.Array;
import def.js.PropertyDescriptor;
import def.rxjs.rxjs.Observable;
import static jsweet.util.Lang.typeof;

@Component(//
		selector = "command", //
		templateUrl = "./ng/command.component.html", styleUrls = { "./ng/command.component.css" })

public class CommandComponent {
	public String command;

	public Array<CommandResponse> responseArray = new Array<>();

	public CommandResponse lastResponse;

	String url = "/web/cmd/";

	Http http;

	LoggerService log;

	public CommandComponent(Http http, LoggerService log) {
		this.http = http;
		this.log = log;
	}

	public void onResponse(long requestTime, String command, Object json) {
		//
		PropertyDescriptor typeP = def.js.Object.getOwnPropertyDescriptor(json, "type");
		String type = (String) typeP.value;

		Array<def.js.String> names = def.js.Object.getOwnPropertyNames(json);

		CommandResponse rt = new CommandResponse(requestTime, command, json);

		this.responseArray.unshift(rt);//

	}

	public void onButtonClick() {
		Globals.console.log("on Button Click,command:" + this.command);
		String command = this.command;
		long requestTime = System.currentTimeMillis();
		Observable<Response> ores = http.post(url, command);
		ores.toPromise().thenOnFulfilledFunction(new Function<Response, Object>() {

			@Override
			public Object apply(Response t) {
				Object json = t.json();

				CommandComponent.this.onResponse(requestTime, command, json);

				log.debug("post response:" + json);
				return null;
			}
		}).catchOnRejectedFunction(new Function<Object, Object>() {

			@Override
			public Object apply(Object t) {
				log.debug("catchOnRejectedFunction,command:" + command);
				return null;
			}
		});

	}
	public boolean showChart() {
		return true;
	}

	public boolean isNumber(def.js.Object json) {
		String type = typeof(json);
		log.debug("typeof,json:" + json + ",is:" + type);
		return type.equals("number");
	}
}

File: dhandho-jsclient/src/app/command/command.component.ts
import { Component } from '@angular/core';
import { HttpModule, Http, Response } from '@angular/http';
import { CommandResponse } from './command-response';
import { LoggerService } from '../service/logger.service';
import { Observable } from 'rxjs/Observable';

@Component( {
    selector: 'app-command',
    templateUrl: './command.component.html',
    styleUrls: ['./command.component.css']
} )
export class CommandComponent {

    public title: string;

    public command: string;

    public responseArray: Array<CommandResponse> = <any>( new Array<any>() );

    public lastResponse: CommandResponse;

    url = '/web/cmd/';

    http: Http;

    log: LoggerService;

    public constructor( http: Http, log: LoggerService ) {
        this.title = 'Command Component';
        if ( this.command === undefined ) { this.command = null; }
        if ( this.lastResponse === undefined ) { this.lastResponse = null; }
        if ( this.http === undefined ) { this.http = null; }
        if ( this.log === undefined ) { this.log = null; }
        this.http = http;
        this.log = log;
    }

    public onResponse( requestTime: number, command: string, json: any ) {
        const typeP: PropertyDescriptor = Object.getOwnPropertyDescriptor( json, 'type' );
        const type: string = <string>typeP.value;
        const names: Array<string> = Object.getOwnPropertyNames( json );
        const rt: CommandResponse = new CommandResponse( requestTime, command, json );
        this.responseArray.unshift( rt );
    }

    public onButtonClick() {
        // console.log( "on Button Click,command:" + this.command );
        const command: string = this.command;
        const requestTime: number = /* currentTimeMillis */Date.now();
        const ores = this.http.post( this.url, command );

        ores.toPromise().then(( value: Response ) => {
            let text1: string = value.toString();
            let text: string = value.text();
            console.log(value);
            const json: any = JSON.parse( text1 );
            this.onResponse( requestTime, command, json );
            this.log.debug( 'post response:' + json );
            return null;
        } ).catch(( t: any ) => {
            this.log.debug( 'catchOnRejectedFunction,command:' + command );
            return null;
        } );
    }

    public showChart(): boolean {
        return true;
    }

    public isNumber( json: Object ): boolean {
        const type: string = typeof json;
        this.log.debug( 'typeof,json:' + json + ',is:' + type );
        return /* equals */( <any>( ( o1: any, o2: any ) => {
            if ( o1 && o1.equals ) {
                return o1.equals( o2 );
            } else {
                return o1 === o2;
            }
        } )( type, 'number' ) );
    }
}
